### PR TITLE
Keep IPCW results in the list column format predicted by the predict() methods

### PR DESCRIPTION
**Pull request by topepo**
The current code unnests the data. This is a problem if we also are using static predictions (like the predicted event time). 

This PR keeps the list format and adds the required columns to each tibble within the list.

### CODE REVIEW

File: R/ipcw.R [hunk 1]
```diff
@@ -24,9 +24,8 @@ trunc_probs <- function(probs, trunc = 0.01) {
   }
   # will still propagate nulls:
   eval_time <- eval_time[!is.na(eval_time)]
-  eval_time <- unique(eval_time)
```
Comment (topepo) [inline R/ipcw.R:27]: These changes are so that we don't change the order of eval times.

File: R/ipcw.R [hunk 2]
```diff
@@ -102,24 +98,28 @@ graf_weight_time <- function(surv_obj, eval_time, rows = NULL, eps = 10^-10) {
...
-#' @param predictors Not currently used. A potential future slot for models with
-#' informative censoring based on columns in `data`.
-#' @param rows An optional integer vector with length equal to the number of
-#' rows in `data` that is used to index the original data. The default is to
-#' use a fresh index on data (i.e. `1:nrow(data)`).
-#' @param eval_time A vector of finite, non-negative times at which to
-#' compute the probability of censoring and the corresponding weights.
+#' curve). This is an internal function.
+#'
+#' @param predictions A data frame with a column containing a [survival::Surv()]
```
Comment (topepo) [inline R/ipcw.R:103]: We now expect the format to be the same as what we produce via `predict()`. It's an internal function, so I don't mind being extra prescriptive about it.

File: R/ipcw.R [hunk 3]
```diff
@@ -102,24 +98,28 @@ graf_weight_time <- function(surv_obj, eval_time, rows = NULL, eps = 10^-10) {
...
 #' very large weight values.
 #' @param eps A small value that is subtracted from the evaluation time when
 #' computing the censoring probabilities. See Details below.
-#' @return A tibble with columns `.row`, `eval_time`, `.prob_cens` (the
-#' probability of being censored just prior to the evaluation time), and
-#' `.weight_cens` (the inverse probability of censoring weight).
+#' @return The same data are returned with the `pred` tibbles containing
+#' several new columns:
+#'
+#' - `.weight_time`: the time at which the inverse censoring probability weights
```
Comment (topepo) [inline R/ipcw.R:117]: I made these names a little more verbose.

**Review by hfrick — COMMENTED** [review]
I like that it just expects the output of `predict()`, that seems sensible. I fell down a rabbit hole a little bit while looking at this code so this review also contains elements that are likely out of scope for this PR. I wanted to jot them down though so we don't loose it.

Specifically on the unnesting/this PR: 

The `.filter_eval_time()` functions doesn't get used anymore now which means infinite time points can get through to the prodlim function which returns the probabilities of being censored (which makes it error). I supposed `add_graf_weights_vec()` would be a good place to doctor with the `eval_time` values.

Generally:

We should look a consistent handling of "non-standard" values for `eval_time`, i.e. `-Inf`, generally `< 0`, `Inf` (and `NA`). Most predictions from censored are less restrictive than what's outlined in `.filter_eval_time()` which I think is nice. survival and prodLim work (at least in parts) with negative values for time so I don't think it's such a big philosophical stretch to get to infinite values. Regardless, if we decide to restrict possible values for `eval_time` rather than pad results appropriately, we should at least warn when we do that.

The other thing I noticed is that we are slightly modifying the censoring probabilities in two places:

- `predict.censoring_model_reverse_km()` adds an epsilon to the probabilities, the amount is data-derived, not set via an argument

and then, after that:
- `trunc_probs()` prevents very small probs, ie adds enough to make them at least `trunc`

So depending how the data is inside of `predict.censoring_model_reverse_km()` we might be changing it to more than `trunc`, so we might want to make the amount an argument to `predict.censoring_model_reverse_km()` to give us control of the final/total amount.

File: R/ipcw.R [hunk 4]
```diff
@@ -185,49 +185,72 @@ graf_weight_time <- function(surv_obj, eval_time, rows = NULL, eps = 10^-10) {
...
-    rlang::warn("The 'predictors' argument to the survival weighting function is not currently used.", call = FALSE)
+  truth <- .find_surv_col(predictions)
+  .check_censored_right(predictions[[truth]])
+  .check_pred_col(predictions)
+
+  if (!is.null(cens_predictors)) {
+    msg <- "The 'cens_predictors' argument to the survival weighting function is not currently used."
+    rlang::warn(msg)
   }
-  eval_time <- .filter_eval_time(eval_time)
```
[the remark this answers is not in the document]
Comment (topepo) [reply R/ipcw.R:212]: I'll add that back in (in a way that keeps the order of time points the same.

[the remark this answers is not in the document]
Comment (topepo) [reply R/ipcw.R:212]: I've added it back but in `predict_survival()` and `predict_hazard()`. Better to get it when it is initially passed.

[the remark this answers is not in the document]
Comment (topepo) [reply R/ipcw.R:212]: We can talk next week about its restrictiveness and the epsilon and other items.

[the remark this answers is not in the document]
Comment (hfrick) [reply R/ipcw.R:212]: Putting it in `predict_survival()` means that those values are not reaching the corresponding computation functions in censored anymore. It's not a heavily advertised aspect of censored but it's there for many engines. I'll check how visibly it's documented.

**Pull request merged**